### PR TITLE
Add dip-12 Coin Trade Flow

### PR DESCRIPTION
**Pull request by LuZhang-Lou**
Please see motivation and spec details in the doc itself.
Note this PR Intentionally only includes the on chain approach.

### CODE REVIEW

File: dips/dip-12.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+```
+enum Metadata {
+  Undefined,
+  GeneralMetadata(GeneralMetadata),
+  TravelRuleMetadata(TravelRuleMetadata),
+  UnstructuredBytesMetadata(UnstructuredBytesMetadata),
+  RefundMetadata(RefundMetadat),
+  CoinTradeMetadata(CoinTradeMetadata),
+  /// and other types of metadatas
+}
```
Comment (davidiw) [inline dips/dip-12.md:31]: should really figure out a more reasonable way of communicating this without copying around this ever increasing blob....

**Review by davidiw — APPROVED** [review]
good as a first draft

File: dips/dip-12.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+created: 3/5/2021
+---
+# Summary and Motivation
+This document proposes a protocol standard for coin trade flows in the Diem Network. The flow is generic for coin purchase and coin sell transactions involving one or more Designated Dealers (DD). It supports a single trade settlement and batch settlement.
+
+A Blockchain coin trade is the process where an entity, for example a Virtual Asset Service Provider (VASP), transacts with a DD to purchase or sell coins. Beyond on-chain p2p transactions, a coin trade usually involves several stages ranging from trade initialization, fiat currency (or other types of cryptocurrency) debitted, to on-chain transfer. The first few steps may vary from entities to entities. However, we see the possibilities and benefits to provide a general specification for the on-chain phase, in pursuit of interoperability among Diem Blockchain participants.
+
+# Terminology
+* trade_id
+  * a unified trade identifier used over the course, a UTF-8 string, max length 255 bytes. It is agreed upon, by both entities, during the trade initialization and negotiation phase. It should be buyer-seller pair-wise unique.
```
Comment (davidiw) [inline dips/dip-12.md:17]: probably make a single line:
```
* ``trade_id`` - ...
```

File: dips/dip-12.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+dip: 12
+title: Blockchain Coin Trade Flow
+authors: Lu Zhang (@LuZhang-Lou)
+dip manager: David Wolinsky (@davidiw)
+status: Draft
+type: Informational
+created: 3/5/2021
+---
+# Summary and Motivation
+This document proposes a protocol standard for coin trade flows in the Diem Network. The flow is generic for coin purchase and coin sell transactions involving one or more Designated Dealers (DD). It supports a single trade settlement and batch settlement.
```
Comment (davidiw) [inline dips/dip-12.md:11]: "or more DDs?" how? considering the nature of this doc, I would just say:

This document proposes a standardized method for referring to off-chain purchases and sales of coins on-chain within Diem transaction metadata.

File: dips/dip-12.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+A Blockchain coin trade is the process where an entity, for example a Virtual Asset Service Provider (VASP), transacts with a DD to purchase or sell coins. Beyond on-chain p2p transactions, a coin trade usually involves several stages ranging from trade initialization, fiat currency (or other types of cryptocurrency) debitted, to on-chain transfer. The first few steps may vary from entities to entities. However, we see the possibilities and benefits to provide a general specification for the on-chain phase, in pursuit of interoperability among Diem Blockchain participants.
```
Comment (davidiw) [inline dips/dip-12.md:13]: always like bullets.

I might suggest we state:

The processes for off-chain transactions are typically unique to the parties involved, however, the on-chain transaction can likely be generalized to pointing to a unique identifier established during the coin trade.

File: dips/dip-12.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+/// Transaction metadata format for coin trades (purchases/sells)
+struct CoinTradeMetadataV0 {
+    /// A nonempty list of trade_ids this transaction wants to settle,
+    trade_ids: Vec<String>,
+}
+```
+
+
+`CoinTradeMetadata` holds a list of trade_ids that the coin transfer wants to settle. Refer to [Binary Canonical Serialization](https://github.com/diem/bcs#binary-canonical-serialization-bcs) for serialization/deserialization details, or use [Diem Client SDK](https://github.com/diem/client-sdks) if available.
```
Comment (davidiw) [inline dips/dip-12.md:46]: I think this is will or wants to, since the off-chain isn't dictated by this document.

Comment (davidiw) [reply dips/dip-12.md:46]: I don't think DIPs should refer to the SDKs.

File: dips/dip-12.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+metadata = diem_types.Metadata__RefundMetadata(
+    value=diem_types.RefundMetadata__RefundMetadataV0(
+        value=diem_types.RefundMetadataV0(
+            transaction_version=original_transaction_version,
+            reason=diem_types.RefundReason__OtherReason()
+        )
+    )
+)
+bcs_metadata = metadata.bcs_serialize()
+```
```
Comment (davidiw) [inline dips/dip-12.md:67]: I think this belongs in a specification / guide not a DIP. When I get back next week, we can figure out how we start crafting a spec, with samples, and dummy data

File: dips/dip-12.md (hunk 2 above)
Comment (bmwill) [inline dips/dip-12.md:17]: > max length 255 bytes

If this is the case then the code that introduced this in the diem/diem repo should be updated to enforce this invariant.

File: dips/dip-12.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+/// List of supported transaction metadata format versions for coin trade transaction
+enum CoinTradeMetadata {
+    CoinTradeMetadataV0(CoinTradeMetadatav0),
+}
+
+/// Transaction metadata format for coin trades (purchases/sells)
+struct CoinTradeMetadataV0 {
+    /// A nonempty list of trade_ids this transaction wants to settle,
+    trade_ids: Vec<String>,
```
Comment (bmwill) [inline dips/dip-12.md:41]: Is there a limit to number of trade_ids (other than limited by txn size) that this should enforce?

File: dips/dip-12.md (hunk 1 above)
Comment (bmwill) [reply dips/dip-12.md:31]: We should probably be much more explicit, for example instead of copying this list we should state what the serialized format looks like, e.g. what is the variant number that corresponds to this entry

File: dips/dip-12.md (hunk 6 above)
Comment (bmwill) [reply dips/dip-12.md:67]: What is a DIP then? are DIPs not intended to be specifications?

File: dips/dip-12.md (hunk 7 above)
Comment (LuZhang-Lou) [reply dips/dip-12.md:41]: I don't think there's hard limit given that this number should be small in practice (<< 100) . Are you suggesting to set a threshold here for  general guidance?

File: dips/dip-12.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+# Specification
+We propose a new type of on-chain Metadata - `CoinTradeMetadata`.
+
+```
+enum Metadata {
+  Undefined,
+  GeneralMetadata(GeneralMetadata),
+  TravelRuleMetadata(TravelRuleMetadata),
+  UnstructuredBytesMetadata(UnstructuredBytesMetadata),
+  RefundMetadata(RefundMetadat),
```
Comment (prasincs) [inline dips/dip-12.md:28]: typo
```suggestion
  RefundMetadata(RefundMetadata),
```

File: dips/dip-12.md (hunk 7 above)
Comment (prasincs) [reply dips/dip-12.md:41]: A general guidance would be good here. I'd expect most of the requests to have 1 tradeID? Trying to go over 100 likely points to some misunderstanding of how this field is expected to be used.

File: dips/dip-12.md (hunk 2 above)
Comment (prasincs) [inline dips/dip-12.md:17]: > It should be buyer-seller pair-wise unique.

This could be expanded to read more clearly as "It should be unique for every trade between a buyer and a seller" 
Maybe there could be some guidance on example TradeIDs? like uint64 ids or UUIDs, etc?

File: dips/dip-12.md (hunk 5 above)
Comment (LuZhang-Lou) [reply dips/dip-12.md:46]: I'm going to remove the part of "SDK"

File: dips/dip-12.md (hunk 2 above)
Comment (LuZhang-Lou) [reply dips/dip-12.md:17]: this is recommendation/reference only, I don't want to introduce too many constraints/checks into the implementation. Let me rephrase this to be explicit. cc @CapCap

Comment (LuZhang-Lou) [reply dips/dip-12.md:17]: We don't specify the ids here because this type of IDs can vary a lot per DD. The hope is the buyer and sender can leverage the same id over the course and avoid adding another layer of translation

File: dips/dip-12.md (hunk 6 above)
Comment (davidiw) [reply dips/dip-12.md:67]: I mistyped specification in the above. It should only really be in a guide unless we want embrace a single and consistent language for defining protocols.

File: dips/dip-12.md (hunk 7 above)
Comment (LuZhang-Lou) [reply dips/dip-12.md:45]: IMO this should be decided by the involved buyer sand seller. If they want to settle 100 trades in a on-chain transfer, I don't see why we shouldn't allow them

File: dips/dip-12.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+authors: Lu Zhang (@LuZhang-Lou)
+dip manager: David Wolinsky (@davidiw)
+status: Draft
+type: Informational
+created: 3/5/2021
+---
+# Summary and Motivation
+This document proposes a standardized method for coin trade flows in the Diem Network. The flow is generic for coin purchase and coin sell transactions. It supports a single trade settlement and batch settlement.
+
+
```
Comment (aching) [inline dips/dip-12.md:13]: nit: extra line

File: dips/dip-12.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+A Blockchain coin trade is the process where an entity, for example a Virtual Asset Service Provider (VASP), transacts with a DD to purchase or sell coins. Beyond on-chain p2p transactions, a coin trade usually involves several stages, such as
+  1. trade initialization
+  2. fiat currency (or other types of cryptocurrency) debitted
```
Comment (aching) [inline dips/dip-12.md:16]:
```suggestion
  2. fiat currency (or other types of cryptocurrency) debited
```

File: dips/dip-12.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  3. on-chain transfer.
+ The first couple of steps are typically unique to the parties involved. However, we see the possibilities and benefits to generalize the on-chain phase, in pursuit of interoperability among Diem Blockchain participants.
+
+# Terminology
+* trade_id - a unified trade identifier used over the course, a UTF-8 string. The recommended max length is 255 bytes. It is agreed upon, by both entities, during the trade initialization and negotiation phase. It should be unique for every trade between a buyer and a seller.
```
Comment (aching) [inline dips/dip-12.md:21]: Why is a trade_id expected to be a UTF-8 string rather than a u64/u128 for instance?

File: dips/dip-12.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+---
+# Summary and Motivation
+This document proposes a standardized method for coin trade flows in the Diem Network. The flow is generic for coin purchase and coin sell transactions. It supports a single trade settlement and batch settlement.
+
+
+A Blockchain coin trade is the process where an entity, for example a Virtual Asset Service Provider (VASP), transacts with a DD to purchase or sell coins. Beyond on-chain p2p transactions, a coin trade usually involves several stages, such as
+  1. trade initialization
+  2. fiat currency (or other types of cryptocurrency) debitted
+  3. on-chain transfer.
+ The first couple of steps are typically unique to the parties involved. However, we see the possibilities and benefits to generalize the on-chain phase, in pursuit of interoperability among Diem Blockchain participants.
```
Comment (davidiw) [inline dips/dip-12.md:18]: unique -> distinct

don't use first or third-persons in these (I, we)

File: dips/dip-12.md (hunk 11 above)
Comment (davidiw) [reply dips/dip-12.md:21]: good call, best to say that at trade_id is an identifier defined by the two parties involved and is therefore represented by a UTF-8 string with a recommended maximum length of 255 bytes (or characters?).

File: dips/dip-12.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    CoinTradeMetadataV0(CoinTradeMetadatav0),
+}
+
+/// Transaction metadata format for coin trades (purchases/sells)
+struct CoinTradeMetadataV0 {
+    /// A nonempty list of trade_ids this transaction wants to settle
+    trade_ids: Vec<String>,
+}
+```
+
```
Comment (davidiw) [inline dips/dip-12.md:48]: remember to strip extra lines

**Review by davidiw — COMMENTED** [review]
any idea why it is having issues with integration tests? note that the repo branch now is main instead of master

File: dips/dip-12.md (hunk 11 above)
Comment (LuZhang-Lou) [reply dips/dip-12.md:21]: @aching we've seen non-numeric trade id from some partner. Usually this id is the one that is included in the trade confirmation docs, so if we make them numeric then it adds another layer of transformation for them.

File: dips/dip-12.md [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+status: Draft
+type: Informational
+created: 3/5/2021
+---
+# Summary and Motivation
+This document proposes a standardized method for coin trade flows in the Diem Network. The flow is generic for coin purchase and coin sell transactions. It supports a single trade settlement and batch settlement.
+
+
+A Blockchain coin trade is the process where an entity, for example a Virtual Asset Service Provider (VASP), transacts with a DD to purchase or sell coins. Beyond on-chain p2p transactions, a coin trade usually involves several stages, such as
+  1. trade initialization
```
[the remark this answers is not in the document]
Comment (LuZhang-Lou) [reply dips/dip-12.md:15]: what do you mean ? 1 and 2 are both out-of-band.

**Review by davidiw — APPROVED** [review]
